lib/authorization.js

```diff
@@ -101,6 +109,11 @@ const checkLoggedIn = (req, res, next) => {
       }
     },
     indirect: {
+      unauthenticated: (unauthenticatedResponse) => {
```

Comment:
So this is middleware equivalent to `restrictToLoggedIn` but with a custom per-invocation code and/or location?

Reply:
Yes! Anything else to add or change?

Reply:
The only thing that comes to mind is that we should make the naming consistent. If this is `restrictToLoggedIn()` but with custom handler, then we probably should call it `restrictToLoggedInCode()` or `restrictToLoggedInErrorCode()` or something like that.

Reply:
I couldn't come up with a name for `unauthenticated()` interface. Please do Suggest!
I think a generic name will be better since it can be used with any interface. Also `restrictToLoggedIn()` is not explicitly called for other interfaces even though it is checked internally.

Reply:
> I couldn't come up with a name for unauthenticated() interface

If you think about it from a user's perspective, it is middleware that we are calling:

1. Receive a given path `app.get("/some/path",`
2. Pass it through first middleware filter, if it passes...
3. Pass it through second middleware filter, if it passes...
4. Pass it through third middleware filter, if it passes...
etc.

Each security middleware is a filter that acts like a gate. So `restrictToLoggedIn` means, "only let someone through this middleware of they are logged in". We probably could have called it `requireLogIn` or `verifyLoggedIn`, but all is the same idea.

In that vein:

1. Why not just call it `restrictToLoggedInWithErrorCode` or similar? It is long, but is expressive, and function names are cheap, these aren't mainframes, and it doesn't pass to the browser (even if it did, we would minify it which would replace it with something nice and clear like `f` :-) )
2. Why do your test and the `README` call `restrictToLoggedIn` _after_ calling `unauthenticated` or whatever we rename it)? `unauthenticated` already calls `restrictToLoggedIn`.

Reply:
1. `restrictToLoggedInWithErrorCode` sounds good. 
2. I am seeing `unauthenticated` as a gate before reaching one of `restrictTo*` gates. So, I was thinking `unauthenticated` can be used along with all the other filters. i.e, to expect custom code for all the restrictions. Earlier, user could pick one of the filters without explicitly using `restrictToLoggedIn`. Now the user will have inconsistent experience. i.e, user will not use `restrictToLoggedIn` for the default response code and will use `restrictToLoggedInWithErrorCode` for the custom code.

OPTION 1:
for default response code,
app.get("/secure/loggedin",cansec.restrictToLoggedIn,send200);
app.get("/secure/user/:user",cansec.restrictToSelf,send200);
app.get("/secure/roles/admin",cansec.restrictToRoles("admin"),send200);

for custom response code,
app.get("/secure/loggedin", cansec.unauthenticated({code:302,location:"/login"}), cansec.restrictToLoggedIn,send200);
app.get("/secure/user/:user", cansec.unauthenticated({code:302,location:"/login"}), cansec.restrictToSelf,send200);
app.get("/secure/roles/admin", cansec.unauthenticated({code:302,location:"/login"}),cansec.restrictToRoles("admin"),send200);

user will first set authentication rules and then set authorization rules.

OPTION 2:
for default response code,
app.get("/secure/loggedin",cansec.restrictToLoggedIn,send200);
app.get("/secure/user/:user",cansec.restrictToSelf,send200);
app.get("/secure/roles/admin",cansec.restrictToRoles("admin"),send200);

*User **does not** use `restrictToLoggedIn` with `restrictToSelf` or `restrictToRoles`*

for custom response code,
app.get("/secure/loggedin", cansec.restrictToLoggedInWithErrorCode({code:302,location:"/login"}),send200); // I'm using a **different interface** now.
app.get("/secure/user/:user", cansec.restrictToLoggedInWithErrorCode({code:302,location:"/login"}), cansec.restrictToSelf,send200);
app.get("/secure/roles/admin", cansec.restrictToLoggedInWithErrorCode({code:302,location:"/login"}),cansec.restrictToRoles("admin"),send200);

*User **uses** `restrictToLoggedInWithErrorCode` with `restrictToSelf` or `restrictToRoles`*

If we have `restrictToLoggedInWithErrorCode`, does the user expect  `restrictTo*WithErrorCode`?

Having said that, I do agree that `restrictToLoggedIn` is unnecessary when `unauthenticated` is used. A careful user can choose to use only `unauthenticated` instead.

In case of OPTION 2 custom code, `restrictToLoggedInWithErrorCode` and `restrictToSelf` both check of `checkLoggedIn`

Reply:
OK, _now_ I see where you are going. This is why I was confused.

* the comments implied that `unauthenticated` was just setting the error code, but _not_ doing the actual `restrictToLoggedIn` filtering
* the code in the PR had `unauthenticated` actually calling `restrictToLoggedIn`

The former would look something like:

```js
app.get("/secure/loggedin", cansec.unauthenticated({code:401, location: '/login'}),cansec.restrictToLogged,send200);
```

While the latter would look like this:

```js
app.get("/secure/loggedin", cansec.restrictToLoggedWithErrorCode({code:401, location: '/login'}),send200);
```

Is that It? You were suggesting that `unauthenticated` *just* set the unauthorized error code and location for future filters in _this_ request chain?

Reply:
Sorry for the confusion.
Yes, `unauthenticated` has to just set unauthorized code for future filters in this request chain.

Reply:
Ah, **now** I get it. And I really like it. Yes, 100%, `unauthenticated` makes more sense, and `restrictToLoggedInErrorCode` does not.. But then it shouldn't call `restrictToLoggedIn`.